Documentation/Contributors/CodingGuide/README.md

```diff
@@ -0,0 +1,662 @@
+* :speedboat: Avoid multiple `return` statements in small hot functions since V8 will not inline them .
+
+### `options` Parameters
+
+:art: Many Cesium functions take an `options` parameter to support optional parameters, self-documenting code, and forward compatibility.  For example, consider:
+```javascript
+var sphere = new SphereGeometry(10.0, 32, 16, VertexFormat.POSITION_ONLY);
+```
+It is not clear what the numeric values represent and the caller needs to know the order of parameters.  If this took an `options` parameter, it would look like:
+```javascript
+var sphere = new SphereGeometry({
+    radius : 10.0,
+    stackPartitions : 32,
+    slicePartitions : 16,
+    vertexFormat : VertexFormat.POSITION_ONLY
+});
+```
+* :speedboat: Using `{ /* ... */ }` creates an object literal, which is a memory allocation.  Avoid designing functions that use an `options` parameter if the function is likely to be a hot spot; otherwise, callers will have to use a scratch variable (see [below](#result-parameters-and-scratch-variables)) for performance.  Constructor functions are good candidates for `options` parameters since Cesium avoids constructing objects in hot spots.
+
+### Default Parameter Values
+
+If a _sensible_ default exists for a function parameter or class property, don't require the user to provide it.  For example, `height` defaults to zero in `Cartesian3.fromRadians`:
+```javascript
+Cartesian3.fromRadians = function(longitude, latitude, height) {
+    height = defaultValue(height, 0.0);
+    // ...
+};
+```
+* :speedboat: Use `defaultValue` to assign a default value to a local variable unless doing so could cause an unnecessary memory allocation, e.g.,
+```javascript
+this._mapProjection = defaultValue(options.mapProjection, new GeographicProjection());
+```
+is better written as
+```javascript
+this._mapProjection = defined(options.mapProjection) ? options.mapProjection : new GeographicProjection();
+```
+
+Some common sensible defaults are:
+* `height`: `0.0`
+* `ellipsoid`: `Ellipsoid.WGS84`
+* `show`: `true`
+
+### Throwing Exceptions
+
+* Throw Cesium's `DeveloperError` when the user has a coding error.  The most common errors are missing parameters and out of range parameters.  For example:
+```javascript
+Cartesian3.maximumComponent = function(cartesian) {
+    //>>includeStart('debug', pragmas.debug);
+    if (!defined(cartesian)) {
+        throw new DeveloperError('cartesian is required.');
+    }
+    //>>includeEnd('debug');
+
+    return Math.max(cartesian.x, cartesian.y, cartesian.z);
+};
+```
+* Surround code to check for `DeveloperError` in `includeStart`/`includeEnd` comments as shown above so developer error checks can be optimized out of release builds.  Do not include required side-effects inside `includeStart`/`includeEnd`, e.g.,
+```javascript
+Cartesian3.maximumComponent = function(cartesian) {
+    //>>includeStart('debug', pragmas.debug);
+    var c = cartesian;
```

Comment:
I don't think it's immediately clear to readers why this is bad.  Maybe explain in more detail that `var c = cartesian` gets removed at build time and changes the behavior of the code.